mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2015 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2015 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# Copyright 2015 Javier Iniesta <javieria@antiun.com>
+# Copyright 2017 David Vidal <david.vidal@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    @api.constrains('partner_id', 'list_ids')
+    def _check_partner_id_lists_ids(self):
+        for contact in self:
+            if contact.partner_id:
+                other_contact = self.search([
+                    ('partner_id', '=', contact.partner_id.id),
+                    ('id', '!=', contact.id)
+                ])
+                if contact.list_ids & other_contact.mapped('list_ids'):
+                    raise ValidationError(_("Partner already exists in one of "
+                                            "these mailing lists"))
```

Comment:
Isn't it easier to just put an unique constraint on the `partner_id` field? Since now a contact has several lists, there's no need to have more than 1.

Reply:
With that, you are limiting a lot the possibilities, so I prefer to not do it.

Reply:
I think the same as @pedrobaeza , besides, in this way it is guaranteed that a partner is not repeated in a list and at the same time I do not limit things so much